torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -124,10 +124,10 @@ def hook_with_zero_fn(
             overlap_info.bucket_index_to_bucket[bucket_index] = bucket
             overlap_info.bucket_index_to_future[bucket_index] = fut
 
-        num_buckets = len(overlap_info.params_per_bucket)
         # NOTE: The implementation from this point forward assumes that the
         # buckets are indexed incrementally starting from 0 in the order of
         # their autograd hooks firing
+        num_buckets = len(overlap_info.params_per_bucket)
         is_last_bucket = bucket_index == num_buckets - 1
```

Comment:
I believe that this `is_last_bucket` is not the same as `GradBucket.is_last()`. The former refers to the order of the all-reduce begins, while the latter refers to the order of the [all-reduce completions](https://github.com/pytorch/pytorch/blob/b03b45afd93b9686418d5d65997d013af2db4038/torch/csrc/distributed/c10d/comm.hpp#L65).

As a sanity check, I have verified that making the change breaks the implementation.